### ISSUE TITLE
Tab Preview Panel Is Blank

### ISSUE THREAD

**Issue by kaos-55** [easonwong-de/Tab-Preview-On-Hover#1]
Hi, I recently saw your post on reddit about this extension so I decided to try the extension, however when I install it and follow the steps to add the css file, it only shows a black box and not the tab thumbnail. I tried it in the nightly and stable version in a clean install and in neither works.
Attached an image of what it looks like after installation.
![firefox-tab-preview](https://user-images.githubusercontent.com/86812371/231466951-d05a6832-9e9e-4a84-adc3-982bc73a6ca5.png)

**Comment by easonwong-de** [comment]
Hello @kaos-55 ! I’m so grateful that you tried this experimental add-on & CSS theme pack. I’ll do some testing and get back to you as soon as possible.

**Comment by easonwong-de** [comment]
@kaos-55 I tested multiple times on both macOS and Windows 11, but sadly couldn’t replicate the problem. Could you go to `about:debugging#/runtime/this-firefox`, find `Tab Preview On Hover` and click “Inspect”? If the add-on doesn’t work as expected, there might be error messages there.

**Comment by kaos-55** [comment]
Sure here is a image of Tab Preview Inspect
![firefox-debug](https://user-images.githubusercontent.com/86812371/231642360-658e5349-1ab3-4965-a4d0-8ea706073e17.png)
Apparently it does not report any errors. I don't know if it has to do with the operating system, I'm using Windows 10, but I'm going to try it on Linux to see if I run into the same error.

**Comment by easonwong-de** [comment]
@kaos-55 This is truly intriguing. I won’t suggest it’s related to operating system any more, as there isn’t that much of a difference between Windows 10 and 11 in regard to Firefox Add-on.

By the way, I’ll appreciate a lot if you would open the “Console” tab in the Developer Tools, as error messages might be there. As far as I can tell, the CSS theme has been installed and works as expected, which means the problem should be on the add-on’s side.

EDIT: I just came up with another way to narrow down the cause of the issue:

1. Open the “Console” tab of the Developer Tools
2. Type in `browser.theme.getCurrent().then(theme => console.log(theme.images.additional_backgrounds))`
3. Expand the output

Ideally, it will look like this:

<img width="726" alt="Screenshot 2023-04-13 at 07 29 18" src="https://user-images.githubusercontent.com/16183548/231662257-58edb7da-4911-49d5-8ba8-5adbddf60ff6.png">

**Comment by kaos-55** [comment]
I also tried in linux, but the black box does not even appear there.
Here is the image of the console, from what I see here it does throw an error.
![extension-debug](https://user-images.githubusercontent.com/86812371/231700100-35498b54-b284-4092-84db-bba1aece21bd.png)

**Comment by easonwong-de** [comment]
It seems that the add-on didn’t send tab thumbnails to the backend. Maybe you want to check if you are browsing in “Private Window” but haven’t allowed the add-on to run in Private Window.

For the Linux part, the CSS theme probably hasn’t been successfully installed.

**Comment by kaos-55** [comment]
I tried in a private window and is quite strange, in private window the extension does work, but in a normal window it does not.
![extension-debug-2](https://user-images.githubusercontent.com/86812371/231808374-65e7a686-4e58-4ce4-937b-d8e3ec2c407c.png)
Using the debug tool in the private window does not give me any errors.
![extension-debug-3](https://user-images.githubusercontent.com/86812371/231810996-648cc042-845d-4cde-a01e-5146370038c8.png)

**Comment by easonwong-de** [comment]
Haha, at least it works to some extent, although ideally it should work in both modes.

**Comment by CennoxX** [comment]
I have the same problem on one pc, on the other with nearly the same setup it works. Works on private window.

**Comment by easonwong-de** [comment]
> I have the same problem on one pc, on the other with nearly the same setup it works. Works on private window.

Does that mean on one set-up the tab preview works, but on the other it works only in private window?

**Comment by CennoxX** [comment]
> Does that mean on one set-up the tab preview works, but on the other it works only in private window?

Yes, that's what I meant.

On the second computer it works now, I think I know what the problem was. It works only after I have gone through all the tabs once (or reloaded all the tabs). Until then, images = null. Because I pinned many tabs, this error occurred quite often.

**Comment by kaos-55** [comment]
Holy Sh*t is working after reloading the tabs again. 
![firefox-tab-preview-working](https://user-images.githubusercontent.com/86812371/233119617-ed263732-3e16-46ff-8a51-a4574fe77ef7.png)
However when you close the browser and open it again you have to reload all the tabs again for it to work. I don't know if there is a solution to avoid reloading the tabs every time the browser is started.

**Comment by easonwong-de** [comment]
@kaos-55 @CennoxX I think I know what could’ve caused the issue. Is there any way that I can try to replicate the bug?

**Comment by kaos-55** [comment]
These are the steps I did to install the extension:
-Clean install Firefox
-Go to about:config and turn toolkit.legacyUserProfileCustomizations.stylesheets to true.
-CSS installation
-Reboot firefox
-Install extension
-Reboot firefox
Then confirm that the black box appears but not the thumbnail.
All this under Windows 10.

**Comment by CennoxX** [comment]
I have to open some pages, close Firefox, reopen Firefox, take a look at the tab previews, they are black (and `browser.theme.getCurrent().then(theme => console.log(theme.images))` is null) until I opened them all, then I can see them (and theme.images is set). The first option at the Firefox settings (about:preferences) `Open previous windows and tabs` has to be turned on.

**Comment by easonwong-de** [comment]
@kaos-55 @CennoxX Thank for the testing! It helps a lot. I’ll fix the bug in short.

**Comment by XeonG** [comment]
Works fine on 113.0b9 (64-bit) Developer Win10

These two share a Firefox profile....
Does not work on 111.0.1 (64-bit) Win10  
Does not work on 112.0.2 (64-bit) Crapple Mac OS

The addon is synced.. however both have the toolkit.legacyUserProfileCustomizations.stylesheets to true and the style sheet is in the /chrome/userstyles/tab-preview.css folder (works fine on Developer one).... 
Doesn't work in private either.. not sure.. 

```
 browser.theme.getCurrent().then(theme => console.log(theme.images.additional_backgrounds))
13:45:43.877
Uncaught ReferenceError: browser is not defined
    <anonymous> debugger eval code:1
debugger eval code:1:1
```

**Comment by easonwong-de** [comment]
@XeonG As long as preview box shows up, the CSS theme is doing great. So I suppose it’s also the Add-on’s problem.

**Comment by XeonG** [comment]
yeah the css preview shows on both
Does not work on 111.0.1 (64-bit) Win10
Does not work on 112.0.2 (64-bit) Crapple Mac OS

just blank, shame any ideas? I suppose I could switch to the Developer... I just didn't cus those morons at mozilla keep breaking better addons like tabmixplus etc so wanted something more stable but thisi is a good addon to come.. isn't enough bad words under the sun for those running mozilla and all the addon stuff they've broken that used to make firefox great.

**Comment by easonwong-de** [comment]
@XeonG @CennoxX @kaos-55 New version v0.0.2 is now available. I hope the bug is addressed… Any feedback is sincerely appreciated!

**Comment by XeonG** [comment]
ah I switched to developer branch and it's working but hopefully someone else can test it and probably go back.. getting annoyed with multile firefox portables editions all using the same taskbar icon :)

as suggestion for features I hope you add in some options, currently it's just css styling.. but its limited to making changes and restarting.. .some changes I'd like to have are a modifier key like [shift], [alt] etc when held down and mouse is over the tabbar.. it shows all the tab previews (currently can modify the css to do this but it happens when the mouse is over the tabbar instead of just over an actual tab... and also some setting to pass to the css the actual tab width.. so the tab preview width matches the same as the tab width (I don't know how to achieve this currently its really annoying),... with the hover tab preview having a wider width than the other tabs previews so other tabs are just the actual width of the respective tabs..    I coudl make another ticket and show some examples.. .but if the addon could have more options in the browser with a custom css override bit that takes effect immiedilaly it would be nice.. also like color select to use on the shadow/outline of the tab ..well those some feature  ideas.  also showing the current tab title text below the preview tab would be useful to see.

**Comment by kaos-55** [comment]
I have tested the new version and the problem has been fixed but at the cost of losing the option to keep tabs suspended until they are selected.
Unfortunately it interferes with one of the extensions I use (auto tab discard) making it unusable.

**Comment by easonwong-de** [comment]
@XeonG Hi, I’m glad things are working out on your end 😃. Thanks for the suggestions and here are some of my thought. You are also welcomed to open new issues or make a fork.

- **Showing all tab previews when the mouse is over the tab bar and a modifier key is pressed**

The showing or hiding of the preview panels totally relies on the CSS theme. Although the add-on can theoretically detect a modifier key being pressed, it couldn’t control the preview panels.

- **Adapting the tab preview width to match the actual tab width**

The preview panels are essentially a pseudo-element of the tab, with `background-image` as the tab preview itself. This means the preview panels can’t adjust its size based on the image. To achieve that, an overhaul of the CSS theme is required.

- **More options to change the look**

For now, there’s no way to change CSS theme without a reboot. Further updates to the CSS theme might come to adjust the look of the preview panels.

- **Showing the current tab title text below the tab preview**

This is tracked in #2

**Comment by easonwong-de** [comment]
@kaos-55 I’m terribly sorry that the new update caused you inconveniences because I didn’t consider compatibility issues with other add-ons. I’ll track this problem in #3. Once again, thanks for the feedback!

Since the “preview panel being blank” bug is addressed, I’m closing this issue now. If the bug happens again, feel free to reöpen it.
